### PR TITLE
Models

### PR DESCRIPTION
**Pull request by mcrd25**
Our core Models and associated tests implemented for Rails 5 Facebook Clone. :)

### CODE REVIEW

**Comment by mcrd25** [comment]
Actually, active friends are those that actively add him and passive
friends are those that he actively added.


Other than that, your comments are noted and would be addressed asap.

☺😊

File: fbclone/app/models/user.rb [hunk 1]
```diff
@@ -1,6 +1,82 @@
...
 class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  before_save { [email.downcase!, first_name.capitalize!, last_name.capitalize!] }
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+  validates :first_name, presence: true, length: { maximum: 35 }
+  validates :last_name, presence: true, length: { maximum: 35 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true,  length: { in: 3..255 }
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: Are you saying it's not important to validate email length at all or not necessary at database level? Is this only or better validated via UI or integration tests?

File: fbclone/app/models/user.rb [hunk 2]
```diff
@@ -1,6 +1,82 @@
...
+# Indexes
+#
+#  index_users_on_email                 (email) UNIQUE
+#  index_users_on_reset_password_token  (reset_password_token) UNIQUE
+#
+
 class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  before_save { [email.downcase!, first_name.capitalize!, last_name.capitalize!] }
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: Glad to know devise takes care of this for us :)

File: fbclone/db/schema.rb [hunk 3]
```diff
@@ -10,11 +10,65 @@
...
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["requestee_id"], name: "index_friend_requests_on_requestee_id"
+    t.index ["requester_id", "requestee_id"], name: "index_friend_requests_on_requester_id_and_requestee_id", unique: true
+    t.index ["requester_id"], name: "index_friend_requests_on_requester_id"
+  end
+
+  create_table "friendships", force: :cascade do |t|
+    t.integer "requester_id"
+    t.integer "requestee_id"
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/db/schema.rb]: @dipto0321, Thanks for this pointer. This is how we saw it really, but we'll talk about it. After all, its nice to leverage convention over configuration.

File: fbclone/app/models/user.rb [hunk 4]
```diff
@@ -1,6 +1,82 @@
...
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+  validates :first_name, presence: true, length: { maximum: 35 }
+  validates :last_name, presence: true, length: { maximum: 35 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: @dipto0321, there's no need for validating presence / length of first / last name and/or valid email regex?

File: fbclone/app/models/user.rb [hunk 5]
```diff
@@ -1,6 +1,82 @@
...
+  has_many :active_friendships, foreign_key: :requester_id,
+                                class_name: 'Friendship',
+                                dependent: :destroy
+  has_many :passive_friendships, foreign_key: :requestee_id, 
+                                 class_name: 'Friendship',
+                                 dependent: :destroy
+
+
+  has_many :active_friends, through: :active_friendships, 
+                            source: :requestee
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: Duly noted!

File: fbclone/app/models/user.rb [hunk 6]
```diff
@@ -1,6 +1,82 @@
...
+  has_many :passive_friends, through: :passive_friendships,
+                             source: :requester
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: Duly noted!

File: fbclone/app/models/user.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (dipto0321) [reply fbclone/app/models/user.rb]: `VALID_EMAIL_REGEX `

**Comment by adriaanbd** [comment]
Hi @dipto0321 thanks a lot for the review!

As Maya pointed out, this is how we interpreted *active* vs *passive* terminology:

* Active Friends are the friends that you actively pursued by requesting friendship via friend request.
* Passive Friends are the ones that you passively pursued by accepting friendship via a friend request.

Joe.active_friends should return an array of those friends whom he requested friendship.

> Hi @mcrd25 and @adriaanbd ,
> 
> This project was good. Check out below points point_down
> 
>     * [This ](https://github.com/mcrd25/facebook-clone/blob/bf68dc58d4d3e0a45ff6b35b268a050d19283d6e/fbclone/Gemfile#L47) gem should be a part of also development group. so that you can use it also in development and test.
> 
>     * Check this issue #41
> 
>     * Every model should test the only custom attributes, associations, and validation. Also, if there is any callback method then you should also test that method. You don't need to test any default attributes which coming from rails and some of the gems like devise. So please remove unnecessary test cases from specs.
> 
>     * You don't need to check validations for any `ID` like
> 
> 
> ```ruby
> class Friendship < ApplicationRecord
> 	validates :requester_id, presence: true
> 	validates :requestee_id, presence: true
> 
> 	belongs_to :requester, class_name: 'User'
> 	belongs_to :requestee, class_name: 'User'
> end
> ```
> 
> When you say `belongs_to` means it's connected with a table. So, in that case, you don't need any validations for `requester_id` and `requestee_id`. So, please remove this kind of validation from all the places you added it.
> 
> ```ruby
> validates :notifiable_type, presence: true 
> validates :notifiable_id, presence: true
> ```
> 
> Also, no need.
> #### Friendships
> 
> I tried this
> 
>     * There were two users Leon and Joseph. Leon made a friendship with Joseph. After that Leon tries to see his active friends which should be empty and his passive friends should be Joseph. But I am getting the opposite.
> 
>     * Leon and Joseph can make duplication friendships like this
>       
>       Try to fix that problem.
> 
> 
> After fixing this please send another request to re-review this project.
> 
> Thanks pray

File: fbclone/app/models/user.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (dipto0321) [reply fbclone/app/models/user.rb]: I don't know actually why you need to valid length for an email. But it's okay if you want to or you have any solid reason. Now for making email presence true will take care by devise. So you don't need to put it here.

**Comment by dipto0321** [comment]
Hi @mcrd25 and @adriaanbd ,

So far so good. Successfully accomplished most of the challenges 👏 😄 . Only one minor thing about `email` validation. I think you can successfully remove [this](https://github.com/mcrd25/facebook-clone/blob/53ccf854c90538caeea2ea48a43587c959a307b2/fbclone/app/models/user.rb#L32-L33) if there is no valid reason to put this thing here. If I am not wrong then devise will take care all of those validations except email length. The choice is yours 🙂. You asked me about a better way to stop duplication of friendships. I didn't find anything wrong with your solution. I liked it. So, for now, I can say this the better way because it's so simple and more understandable. Maybe in future when you will get used to more with rails you will figure out. For me it's okay. It's completely based on my opinion. This is from me 👍 for your model. You can merge it into the master.

Thanks 🙏 ☺️ 😌

File: fbclone/app/models/user.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: You're right Dipto, Devise does take care of this for us in the `initializers/devise.rb` wherein it provides for a default email regex but only guarantees presence of only one "@". So we might end up moving our custom regex there. Nice to know that Devise also comes with case sensitive false default for email.

File: fbclone/app/models/user.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (adriaanbd) [reply fbclone/app/models/user.rb]: Keeping maximum length for email is probably what we'll end up doing, as Devise takes care of email presence.

**Comment by adriaanbd** [comment]
Thanks for sharing the knowledge @dipto0321! We'll discuss tomorrow on implementing some minor adjustments to our User.email validation in favor of devise defaults and inserting our custom email regex in `devise.rb`.

**Pull request merged**